vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/pool/SqlConnectionPool.java

```diff
@@ -0,0 +1,245 @@
+import io.vertx.core.net.impl.clientconnection.ConnectResult;
+import io.vertx.core.net.impl.clientconnection.Lease;
+import io.vertx.core.net.impl.pool.ConnectionEventListener;
+import io.vertx.core.net.impl.pool.ConnectionPool;
+import io.vertx.core.net.impl.pool.Connector;
+import io.vertx.core.net.impl.pool.SimpleConnectionPool;
+import io.vertx.sqlclient.PoolOptions;
+import io.vertx.sqlclient.impl.Connection;
+import io.vertx.sqlclient.impl.ConnectionFactory;
+import io.vertx.sqlclient.impl.command.CommandBase;
+import io.vertx.sqlclient.spi.DatabaseMetadata;
+import io.vertx.core.*;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Todo :
+ *
+ * - handle timeout when acquiring a connection
+ * - for per statement pooling, have several physical connection and use the less busy one to avoid head of line blocking effect
+ *
+ * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
+ */
+public class SqlConnectionPool {
+
+  private final ConnectionFactory factory;
+  private final ContextInternal context;
+  private final ConnectionPool<PooledConnection> pool;
+  private final int maxSize;
+  private boolean closed;
+
+  public SqlConnectionPool(ConnectionFactory connector, int maxSize) {
+    this(connector, maxSize, PoolOptions.DEFAULT_MAX_WAIT_QUEUE_SIZE);
+  }
+
+  public SqlConnectionPool(ConnectionFactory connector, int maxSize, int maxWaitQueueSize) {
+    this(connector, null, maxSize, maxWaitQueueSize);
+  }
+
+  public SqlConnectionPool(ConnectionFactory factory, Context context, int maxSize, int maxWaitQueueSize) {
+    Objects.requireNonNull(factory, "No null connector");
+    if (maxSize < 1) {
+      throw new IllegalArgumentException("Pool max size must be > 0");
+    }
+    this.pool = new SimpleConnectionPool<>(connector, maxSize, maxSize, maxWaitQueueSize);
+    this.context = (ContextInternal) context;
+    this.maxSize = maxSize;
+    this.factory = factory;
+  }
+
+  private final Connector<PooledConnection> connector = new Connector<PooledConnection>() {
+    @Override
+    public void connect(EventLoopContext context, ConnectionEventListener listener, Handler<AsyncResult<ConnectResult<PooledConnection>>> handler) {
+      PromiseInternal<Connection> promise = context.promise();
+      factory.connect(promise);
+      promise.future()
+        .map(connection -> {
+          PooledConnection pooled = new PooledConnection(connection, listener);
+          connection.init(pooled);
+          return new ConnectResult<>(pooled, 1, 1);
+        })
+        .onComplete(handler);
+    }
+
+    @Override
+    public boolean isValid(PooledConnection connection) {
+      return true;
+    }
+  };
+
+  public int available() {
+    return maxSize - pool.size();
+  }
+
+  public int size() {
+    return pool.size();
+  }
+
+  public void acquire(ContextInternal context, Handler<AsyncResult<Connection>> waiter) {
+    pool.acquire((EventLoopContext) context, 1, ar -> {
```

Comment:
This cast will throw a `ClassCastException` under `DuplicatedContext` for example it's called in a HTTP Server request handler.

Reply:
alright it should need an integration test to ensure it's fine (and be corrected)